experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,266 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+// TODO: add module-level Rust-doc with examples
+
+use crate::error::Error;
+use crate::impl_const::*;
+use std::marker::PhantomData;
+use zerovec::ZeroVec;
+
+// Enums
+
+/// The width of the elements in the data array of a [`CodePointTrie`].
+/// See UCPTrieValueWidth in ICU4C.
+#[derive(Clone, Copy, PartialEq)]
+pub enum ValueWidthEnum {
+    Bits16 = 0,
+    Bits32 = 1,
+    Bits8 = 2,
+}
+
+/// The type of trie represents whether the trie has an optimization that
+/// would make it small or fast.
+/// See UCPTrieType in ICU4C.
+#[derive(Clone, Copy, PartialEq)]
+pub enum TrieTypeEnum {
+    Fast = 0,
+    Small = 1,
+}
+
+// ValueWidth trait
+
+// AsULE is AsUnalignedLittleEndian, i.e. "allowed in a zerovec"
+pub trait ValueWidth: Copy + zerovec::ule::AsULE {
+    const ENUM_VALUE: ValueWidthEnum;
+    fn cast_to_widest(self) -> u32;
+
+    fn get_code_point_trie_value_width_enum(value_width_int: u8) -> Option<ValueWidthEnum> {
+        match value_width_int {
+            0 => Some(ValueWidthEnum::Bits16),
+            1 => Some(ValueWidthEnum::Bits32),
+            2 => Some(ValueWidthEnum::Bits8),
+            _ => None,
+        }
+    }
+}
+
+impl ValueWidth for u8 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits8;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u16 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits16;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u32 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits32;
+
+    fn cast_to_widest(self) -> u32 {
+        self
+    }
+}
+
+// TrieType trait
+
+pub trait TrieType {
+    const FAST_MAX: u32;
+    const ENUM_VALUE: TrieTypeEnum;
+
+    fn get_code_point_trie_type_enum(trie_type_int: u8) -> Option<TrieTypeEnum> {
+        match trie_type_int {
+            0 => Some(TrieTypeEnum::Fast),
+            1 => Some(TrieTypeEnum::Small),
+            _ => None,
+        }
+    }
+}
+
+pub struct Fast;
+impl TrieType for Fast {
+    const FAST_MAX: u32 = 0xffff;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Fast;
+}
+
+pub struct Small;
+impl TrieType for Small {
+    const FAST_MAX: u32 = 0x0fff;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    header: CodePointTrieHeader,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+pub struct CodePointTrieHeader {
+    pub index_length: u32,
+    pub data_length: u32,
+    pub high_start: u32,
+    pub shifted12_high_start: u16,
+    pub index3_null_offset: u16,
+    pub data_null_offset: u32,
+    pub null_value: u32,
+}
+
+// TODO: add Rust-doc that includes examples
```

Comment:
As discussed in email with @markusicu: Add a function similar to [`UnicodeSet::bmp()`](https://unicode-org.github.io/icu4x-docs/doc/icu_uniset/struct.UnicodeSet.html#method.bmp), perhaps `CodePointTrie::planes()`, that maps every code point to its plan number (0 through 16).  You can then use this trie in your documentation examples.